src/did-change-watched-files.jl

```diff
@@ -0,0 +1,93 @@
+                    kind = WatchKind.Create | WatchKind.Change | WatchKind.Delete),
+            ]))
+end
+
+# For dynamic registrations during development
+# unregister(currently_running, Unregistration(;
+#     id=DID_CHANGE_WATCHED_FILES_REGISTRATION_ID,
+#     method=DID_CHANGE_WATCHED_FILES_REGISTRATION_METHOD))
+# register(currently_running, did_change_watched_files_registration())
+
+function initialize_config!(server::Server)
+    if !isdefined(server.state, :root_path)
+        if JETLS_DEV_MODE
+            @info "server.state.root_path is not defined, skipping add JETLSConfig.toml watcher"
+        end
+        return
+    end
+    config_path = joinpath(server.state.root_path, "JETLSConfig.toml")
+    push!(server.state.config_manager.watching_files, config_path)
+    load_config!(server, config_path) do actual_config, latest_config, key_path, v
+        # at initialization, we can just update the config in both actual and latest configs
+        actual_config[last(key_path)] = v
+        latest_config[last(key_path)] = v
+    end
+end
+
+"""
+Loads the configuration from the specified path into the server's config manager.
+
+If the file does not exist or cannot be parsed, current configuration remains unchanged.
+When there are unknown keys in the config file, send error by `workspace/showMessage`
+and **current configuration is not changed.**
+"""
+function load_config!(on_reload_required, server::Server, path::AbstractString)
+    isfile(path) || return
+    parsed = TOML.tryparsefile(path)
+    parsed isa TOML.ParserError && return # just skip to reduce noise while typing
+    unknown_keys = collect_unmatched_keys(parsed)
+
+    if !isempty(unknown_keys)
+        show_error_message(server, """
+            Configuration file at $path contains unknown keys.
+            unknown keys: $(join(map(x -> join(x, "."), unknown_keys), ", "))
+            """)
+        return
+    end
+
+    merge_config!(on_reload_required,
+                  server.state.config_manager,
+                  parsed)
+end
+
+function handle_file_change!(server::Server, change::FileEvent)
+    changed_path = uri2filepath(change.uri)
+    change_type = change.type
+    is_config_file(server, changed_path) || return
+    if change_type == FileChangeType.Created || change_type == FileChangeType.Changed
+        load_config!(server, changed_path) do _, latest_config, key_path, v
+            k = last(key_path)
+            if latest_config[k] !== v
+                latest_config[k] = v
+                show_warning_message(server, """
+                    Configuration key `$(join(key_path, "."))` was changed.
+                    Please restart the server to apply the changes.
+                    """)
+            end
+        end
+    elseif change_type == FileChangeType.Deleted
+        show_warning_message(server, """
+            $changed_path was deleted.
+            Please restart the server to apply the changes.
+            """)
```

Comment:
How about extending the `ConfigManager` data structure to maintain a mapping of `config_file_path => configs`?
That way, when a configuration file is deleted, it should be possible to delete the configuration held in that mapping and revert to the default.
In the future, when we want to look at multiple configuration files (such as those under `~/.config`), such a mapping will be necessary to implement a configuration that reflects that hierarchy.

Reply:
I agree that your suggested implementation would likely be more robust to future changes. I’ve updated the code accordingly, though I haven’t yet fully verified the implementation details. I’ll let you know once it’s ready for a proper review.

- Each config file’s contents are now stored in a `SortedDict` mapping filename => content. Default settings and command-line arguments are also stored in this dict using special reserved names.
- The dict is sorted by `ConfigFileOrder`, so config values can be retrieved correctly by iterating it.
- Removing a file simply handled deleting its corresponding key from the dict.